primitives/src/crh/poseidon/mod.rs

```diff
@@ -1,427 +1,123 @@
+        idx_cst += P::T;
     }
+    *state = new_state;
+}
 
-    fn poseidon_partial_round(vec_state: &mut Vec<Vec<P::Fr>>, round_cst_idx: &mut usize) {
-
-        // For each of the state vector element position
-        for j in 0..P::T {
-
-            // get the constant associated to state vector element position
-            let rc = P::ROUND_CST[*round_cst_idx];
-
-            // go over each state vector
-            for k in 0..vec_state.len() {
-                vec_state[k][j] += &rc;
-            }
-            *round_cst_idx += 1;
-        }
-
-        // Apply the S-BOX to the first elements of each of the state vector
-        let mut w: Vec<P::Fr> = Vec::new();
-        let mut accum_prod = P::Fr::one();
+impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr=F>> PoseidonHash<F, P> {
 
-        w.push(accum_prod);
-        // Calculate the intermediate partial products
-        for i in 0..vec_state.len() {
-            accum_prod = accum_prod * &vec_state[i][0];
-            w.push(accum_prod);
-        }
-
-        // if the accum_prod is zero, it means that one of the S-Boxes is zero
-        // in that case compute the inverses individually
-        if accum_prod == P::Fr::zero() {
-            for i in 0..(vec_state.len() - 1) {
-                if vec_state[i][0] != P::Fr::zero() {
-                    vec_state[i][0] = vec_state[i][0].inverse().unwrap();
-                }
-            }
-        } else {
-
-            // Calculate the inversion of the products
-            // Use Montgomery simultaneous inversion
-            let mut w_bar = accum_prod.inverse().unwrap();
-
-            // Extract the individual inversions
-            let mut idx: i64 = w.len() as i64 - 2;
-            for i in (0..vec_state.len()).rev() {
-                let vec_1 = vec_state[i][0].clone();
-                vec_state[i][0] = w_bar * &w[idx as usize];
-                w_bar = w_bar * &vec_1;
-                idx -= 1;
-            }
+    #[inline]
+    fn apply_permutation(&mut self) {
+        for (input, state) in self.pending.iter().zip(self.state.iter_mut()) {
+            *state += input;
         }
+        self.state[P::R] += &P::C2;
+        Self::poseidon_perm(&mut self.state);
     }
 
-    fn poseidon_perm_gen(vec_state: &mut Vec<Vec<P::Fr>>) {
-
-        // index that goes over the round constants
-        let mut round_cst_idx: usize = 0;
-
-        // Full rounds
-        for _i in 0..P::R_F {
-            Self::poseidon_full_round(vec_state, &mut round_cst_idx);
-
-            // Perform the matrix mix
-            for i in 0..vec_state.len() {
-                //Self::matrix_mix_short(&mut vec_state[i]);
-                Self::matrix_mix_short(&mut vec_state[i]);
-            }
-
+    #[inline]
+    fn _finalize(&self) -> F {
+        let mut state = self.state.clone();
+        for (input, s) in self.pending.iter().zip(state.iter_mut()) {
```

Comment:
This doesn't handle correctly padding. Create two versions: one for constant size inputs hashing and another one for variable size taking in consideration if the variable size data will always be modulus rate (in such a case there is no need for padding).